lib/Controller/LoginController.php

```diff
@@ -161,6 +162,40 @@ public function customOidc($provider)
         return $this->auth(CustomOpenIDConnect::class, $config, $provider, 'OpenID Connect');
     }
 
+    /**
+     * @PublicPage
+     * @NoCSRFRequired
+     */
+    public function customOauth2($provider)
+    {
+        $config = [];
+        $providers = json_decode($this->config->getAppValue($this->appName, 'custom_oauth2_providers', '[]'), true);
+        if (is_array($providers)) {
+            foreach ($providers as $prov) {
+                if ($prov['name'] === $provider) {
+                    $callbackUrl = $this->urlGenerator->linkToRouteAbsolute($this->appName.'.login.custom_oauth2', ['provider' => $provider]);
+                    $config = [
+                        'callback' => $callbackUrl,
+                        'scope' => $prov['scope'],
+                        'keys' => [
+                            'id'     => $prov['clientId'],
+                            'secret' => $prov['clientSecret'],
+                        ],
+                        'endpoints' => new Data\Collection([
+                            'api_base_url'     => $prov['apiBaseUrl'],
+                            'authorize_url'    => $prov['authorizeUrl'],
+                            'access_token_url' => $prov['tokenUrl'],
+                            'profile_url'      => $prov['profileUrl'],
+                            'profile_fields'   => $prov['profileFields']
```

Comment:
`profile_fields` is not "endpoint".

Reply:
It is - see all the other providers from HybridAuth - e.g. Facebook or LinkedIn. They only have their "fields" hard-coded.